Framework/Muon/src/MuonPreProcess.cpp

```diff
@@ -214,18 +235,68 @@ MatrixWorkspace_sptr MuonPreProcess::applyTimeOffset(MatrixWorkspace_sptr ws,
   }
 }
 
+MatrixWorkspace_sptr
+MuonPreProcess::applyTimeZeroTable(MatrixWorkspace_sptr ws,
+                                   const TableWorkspace_sptr &timeZeroTable) {
+  auto cloneWs = cloneWorkspace(ws);
+  if (timeZeroTable != nullptr) {
+    const auto numSpec = cloneWs->getNumberHistograms();
+    for (auto specNum = 0u; specNum < numSpec; ++specNum) {
+      auto &xData = cloneWs->mutableX(specNum);
+      for (auto &xValue : xData) {
+        API::TableRow row = timeZeroTable->getRow(specNum);
+        xValue -= row.Double(0);
+      }
+    }
+  }
+  return cloneWs;
+}
+
 MatrixWorkspace_sptr MuonPreProcess::applyCropping(MatrixWorkspace_sptr ws,
                                                    const double &xMin,
                                                    const double &xMax) {
   if (xMin != EMPTY_DBL() || xMax != EMPTY_DBL()) {
-    IAlgorithm_sptr crop = createChildAlgorithm("CropWorkspace");
-    crop->setProperty("InputWorkspace", ws);
-    if (xMin != EMPTY_DBL())
-      crop->setProperty("Xmin", xMin);
-    if (xMax != EMPTY_DBL())
-      crop->setProperty("Xmax", xMax);
-    crop->execute();
-    return crop->getProperty("OutputWorkspace");
+    if (!getPropertyValue("TimeZeroTable").empty()) {
```

Comment:
This very long block really should be broken down into multiple methods - any future developers working on this have to not only figure out which `if-else` block their in, but also figure out what individual steps do rather than just have step names like

```
std::pair<double, double> getMinMaxValue() { 
   // lines 262-287
}

const auto [xMin, xMax] = getMinMaxValue();
cropRagged.setProperty(XMax, xMax);
cropRagged.setProperty(XMin, xMin);
...etc.
```